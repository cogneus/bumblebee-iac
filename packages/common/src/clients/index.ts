export * from './db'
export * from './secrets'
export * from './ssm'
export * from './ensure-client'
export * from './types'
export * from './__mocks__'
export * from './destroy-clients'
