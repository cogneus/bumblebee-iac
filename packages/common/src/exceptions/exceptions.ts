import { AppException } from './app-exception.class'
import { ExceptionCodes } from './exception-codes'
import { StatusCodes } from './status-codes'

export const InvalidToken = () =>
  new AppException({
    code: ExceptionCodes.InvalidToken,
    status: StatusCodes.Forbidden,
    message: 'Forbidden',
    source: InvalidToken,
  })

export const InvalidAuthorizartion = () =>
  new AppException({
    code: ExceptionCodes.InvalidAuthorizartion,
    status: StatusCodes.Forbidden,
    message: 'Forbidden',
    source: InvalidAuthorizartion,
  })

export const EmptyRequest = () =>
  new AppException({
    code: ExceptionCodes.EmptyRequest,
    status: StatusCodes.BadRequest,
    message: 'Request data is missing',
    source: EmptyRequest,
  })

export const InvalidService = (service: string) =>
  new AppException({
    code: ExceptionCodes.InvalidService,
    status: StatusCodes.Error,
    message: `Invalid service. ${service}`,
    source: InvalidService,
  })

export const InvalidRequest = (innerMessage: string) =>
  new AppException({
    code: ExceptionCodes.InvalidRequest,
    status: StatusCodes.BadRequest,
    message: `Request data is invalid. ${innerMessage}`,
    source: InvalidRequest,
  })

export const MissingAttributes = (keys: string[]) =>
  new AppException({
    code: ExceptionCodes.MissingAttributes,
    status: StatusCodes.BadRequest,
    message: `The following attributes are mandatory ${keys.join()}`,
    source: MissingAttributes,
  })

export const EnvVarNotFound = (keys: string[]) =>
  new AppException({
    code: ExceptionCodes.EnvVarNotFound,
    status: StatusCodes.Error,
    message: `The following environment variables not found ${keys.join()}`,
    source: EnvVarNotFound,
  })

export const InvalidModeSource = (resourcePrefix: string, eventSourceARN: string | undefined) =>
  new AppException({
    code: ExceptionCodes.InvalidModeSource,
    status: StatusCodes.Error,
    message: `Cannot derive stage from event source. eventSourceARN: ${eventSourceARN} RESOURCE_PREFIX: ${resourcePrefix}`,
    source: InvalidModeSource,
  })
