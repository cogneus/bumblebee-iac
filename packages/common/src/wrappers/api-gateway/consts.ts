/* eslint @typescript-eslint/naming-convention: 0 */
export const APIResponseHeaders = { 'Content-Type': 'application/json' }
