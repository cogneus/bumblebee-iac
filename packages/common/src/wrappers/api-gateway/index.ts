export * from './api-gateway.wrapper'
export * from './types'
