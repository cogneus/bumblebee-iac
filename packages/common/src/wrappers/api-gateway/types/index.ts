export * from './api-gateway-dependencies.interface'
export * from './api-gateway-handler.type'
export * from './api-gateway-wrapper.type'
export * from './api-gateway-auth-checker.type'
export * from './api-gateway-wrapper-params.interface'
