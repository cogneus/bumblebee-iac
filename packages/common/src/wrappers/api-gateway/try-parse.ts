import { InvalidRequest } from '../../exceptions'

export const tryParse = <TPayload>(body: string | undefined | null): TPayload => {
  if (!body) {
    return undefined as TPayload
  }
  try {
    return JSON.parse(body) as TPayload
  } catch (ex) {
    throw InvalidRequest(ex.message)
  }
}
