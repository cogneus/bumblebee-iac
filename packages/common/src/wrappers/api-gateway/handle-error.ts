import { APIGatewayProxyResultV2 } from 'aws-lambda'
import { AppException, StatusCodes } from '../../exceptions'
import { APIResponseHeaders } from './consts'

const forbidenErrorBody = {
  message: 'forbidden',
}
export const handleError = ({
  code,
  message,
  status,
}: Error & AppException): APIGatewayProxyResultV2 => ({
  statusCode: status || StatusCodes.Error,
  headers: APIResponseHeaders,
  body: JSON.stringify(
    status === StatusCodes.Forbidden
      ? forbidenErrorBody
      : {
          message,
          code,
        }
  ),
})
