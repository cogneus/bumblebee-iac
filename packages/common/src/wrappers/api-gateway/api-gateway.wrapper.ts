import 'source-map-support/register'
import { InvalidAuthorizartion, StatusCodes } from '../../exceptions'
import { ServiceContext } from '../../types'
import { createWrapperContext } from '../utils'
import { decodeToken } from '../../auth'
import { handleError } from './handle-error'
import { tryParse } from './try-parse'
import { APIGatewayWrapper, APIGatewayWrapperParams } from './types'
import { destroyClients } from '../../clients'
import { APIResponseHeaders } from './consts'

export const apiGatewayWrapper: APIGatewayWrapper =
  <TPayload, TParams, TToken, TResult>({
    handler,
    getClientOptions,
    authCheck,
    ...contextParams
  }: APIGatewayWrapperParams<TPayload, TParams, TToken, TResult>) =>
  async (event, context: ServiceContext) => {
    const { body, pathParameters, queryStringParameters, headers } = event
    const params = {
      ...(pathParameters as Record<string, any>),
      ...(queryStringParameters as Record<string, any>),
    } as TParams
    const serviceContext = createWrapperContext({
      ...contextParams,
      context,
      logInfo: {
        pathParameters,
        queryStringParameters,
        body,
      },
      clientOptions: getClientOptions(params),
    })
    const { log } = serviceContext

    try {
      const token: TToken = decodeToken(headers)
      if (!authCheck(token, serviceContext)) {
        throw InvalidAuthorizartion()
      }
      log.debug('Process input', event)

      const result = await handler({
        params,
        payload: tryParse<TPayload>(body),
        serviceContext,
      })

      log.debug('Process complete', result as any)
      return {
        body: JSON.stringify(result),
        statusCode: StatusCodes.Success,
        headers: APIResponseHeaders,
      }
    } catch (ex) {
      log.warn(`Process failed, ${ex.message}`)
      log.error(ex)
      return handleError(ex)
    } finally {
      destroyClients(serviceContext.clients)
    }
  }
