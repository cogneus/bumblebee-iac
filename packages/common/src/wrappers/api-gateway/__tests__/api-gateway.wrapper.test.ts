import { APIGatewayProxyEventV2 } from 'aws-lambda'
import { apiGatewayWrapper } from '../api-gateway.wrapper'
import { ClientOptions } from '../../../clients'
import { AppException, StatusCodes } from '../../../exceptions'

const clientOptions: ClientOptions = {
  db: {
    maxAttempts: 3,
    region: 'eu-west-1',
    requestTimeout: 3000,
    tables: {
      test: {
        name: 'test-table-name',
        indices: {
          testIndex: { name: 'test-index-name' },
        },
        itemLimits: { defaultLimit: 20, maxLimit: 100 },
      },
    },
  },
}

const events: Partial<APIGatewayProxyEventV2>[] = [
  {
    headers: {
      authorization:
        'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c',
    },
    pathParameters: {
      testPathParam: 'test-path-param',
    },
    queryStringParameters: {
      testQueryParam: 'test-query-param',
    },
    body: '{"test": "body"}',
  },
]

const invalidEvents: {
  event: Partial<APIGatewayProxyEventV2>
  error?: AppException
}[] = [
  {
    error: {
      name: 'NotFound',
      code: 'NotFound',
      message: 'NotFound',
      status: StatusCodes.NotFound,
    },
    event: {
      headers: {
        authorization:
          'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJyaWdodHMiOlsiV1JJVEUiXX0.0_slzFK8MqvPbpToNcRqPBjrDjUL_b6XbgqnMNK4Pe4',
      },
      pathParameters: {
        testPathParam: 'test-path-param',
      },
      queryStringParameters: {
        testQueryParam: 'test-query-param',
      },
      body: '{"test": "body"}',
    },
  },
  {
    error: {
      name: 'Error',
      code: 'Error',
      message: 'Error',
      status: StatusCodes.Error,
    },
    event: {
      headers: {
        authorization:
          'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJyaWdodHMiOlsiV1JJVEUiXX0.0_slzFK8MqvPbpToNcRqPBjrDjUL_b6XbgqnMNK4Pe4',
      },
      pathParameters: {
        testPathParam: 'test-path-param',
      },
      queryStringParameters: {
        testQueryParam: 'test-query-param',
      },
      body: '{"test": "body"@',
    },
  },
  {
    error: {
      name: 'Forbidden',
      code: 'Forbidden',
      message: 'Forbidden',
      status: StatusCodes.Forbidden,
    },
    event: {
      headers: {
        authorization:
          'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJyaWdodHMiOlsiV1JJVEUiXX0.0_slzFK8MqvPbpToNcRqPBjrDjUL_b6XbgqnMNK4Pe4',
      },
      pathParameters: {
        testPathParam: 'test-path-param',
      },
      queryStringParameters: {
        testQueryParam: 'test-query-param',
      },
      body: '{"test": "body"}',
    },
  },
  {
    error: {
      name: 'BadRequest',
      code: 'BadRequest',
      message: 'BadRequest',
      status: StatusCodes.BadRequest,
    },
    event: {
      headers: {
        authorization:
          'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJyaWdodHMiOlsiV1JJVEUiXX0.0_slzFK8MqvPbpToNcRqPBjrDjUL_b6XbgqnMNK4Pe4',
      },
      pathParameters: {
        testPathParam: 'test-path-param',
      },
      queryStringParameters: {
        testQueryParam: 'test-query-param',
      },
      body: '{"test": "body"}',
    },
  },
  {
    event: {
      headers: {
        authorization:
          'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJyaWdodHMiOlsiV1JJVEUiXX0.0_slzFK8MqvPbpToNcRqPBjrDjUL_b6XbgqnMNK4Pe4',
      },
      pathParameters: {
        testPathParam: 'test-path-param',
      },
      queryStringParameters: {
        testQueryParam: 'test-query-param',
      },
      body: '{"test": "body"',
    },
  },
  {
    event: {
      headers: {
        authorization:
          'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJyaWdodHMiOlsiUkVBRCJdfQ.Qhhxqxb7HqVcdSUSnmntHi1VkMGDkD0o3VoWgzvy-H8',
      },
      pathParameters: {
        testPathParam: 'test-path-param',
      },
      queryStringParameters: {
        testQueryParam: 'test-query-param',
      },
      body: '{"test": "body"}',
    },
  },
]

describe('api gateway wrapper tests', () => {
  describe('should handle response, returns', () => {
    it.each(events)('results matching snapshot', async (event) => {
      jest.resetAllMocks()
      const results: any[] = []
      const wrapper = apiGatewayWrapper({
        handler: async ({ serviceContext, params, payload }) => {
          results.push({
            params,
            payload,
            serviceContext,
          })
          return {
            body: '{}',
            statusCode: 200,
          }
        },
        region: 'eu-west-1',
        env: 'dev',
        logName: 'wrapper-test',
        serviceName: 'wrapper-test',
        getClientOptions: () => clientOptions,
        authCheck: () => {
          return true
        },
      })

      const cb: any = null
      await wrapper(event as APIGatewayProxyEventV2, {} as any, cb)
      expect(results).toMatchSnapshot()
    })
  })

  describe('should throw Error, returns', () => {
    it.each(invalidEvents)('results matching snapshot', async ({ event, error }) => {
      jest.resetAllMocks()
      const wrapper = apiGatewayWrapper({
        handler: async () => {
          if (error) {
            throw error
          }
          return {
            body: '{}',
            statusCode: 200,
          }
        },
        region: 'eu-west-1',
        env: 'dev',
        logName: 'wrapper-test',
        serviceName: 'wrapper-test',
        getClientOptions: () => clientOptions,
        authCheck: (token: any) => token?.rights.includes('WRITE') || false,
      })

      const cb: any = null
      const result = await wrapper(event as APIGatewayProxyEventV2, {} as any, cb)
      expect(result).toMatchSnapshot()
    })
  })
})
