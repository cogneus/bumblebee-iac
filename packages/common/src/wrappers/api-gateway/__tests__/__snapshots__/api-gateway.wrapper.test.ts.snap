// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api gateway wrapper tests should handle response, returns results matching snapshot 1`] = `
[
  {
    "params": {
      "testPathParam": "test-path-param",
      "testQueryParam": "test-query-param",
    },
    "payload": {
      "test": "body",
    },
    "serviceContext": {
      "clientOptions": {
        "db": {
          "maxAttempts": 3,
          "region": "eu-west-1",
          "requestTimeout": 3000,
          "tables": {
            "test": {
              "indices": {
                "testIndex": {
                  "name": "test-index-name",
                },
              },
              "itemLimits": {
                "defaultLimit": 20,
                "maxLimit": 100,
              },
              "name": "test-table-name",
            },
          },
        },
      },
      "env": "dev",
      "log": ConsoleLogger {
        "config": {
          "formatter": [Function],
          "includeSrc": true,
          "level": undefined,
          "levels": undefined,
        },
        "defaultEventCode": undefined,
        "formatter": [Function],
        "levels": {
          "debug": 2,
          "error": 5,
          "info": 3,
          "trace": 1,
          "warn": 4,
        },
        "minLevel": 3,
        "resolveLevel": [Function],
        "scope": {
          "body": "{"test": "body"}",
          "env": "dev",
          "name": "wrapper-test",
          "pathParameters": {
            "testPathParam": "test-path-param",
          },
          "queryStringParameters": {
            "testQueryParam": "test-query-param",
          },
          "region": "eu-west-1",
          "serviceName": "wrapper-test",
        },
        "transformer": undefined,
      },
      "region": "eu-west-1",
    },
  },
]
`;

exports[`api gateway wrapper tests should throw Error, returns results matching snapshot 1`] = `
{
  "body": "{"message":"NotFound","code":"NotFound"}",
  "headers": {
    "Content-Type": "application/json",
  },
  "statusCode": 404,
}
`;

exports[`api gateway wrapper tests should throw Error, returns results matching snapshot 2`] = `
{
  "body": "{"message":"Request data is invalid. Unexpected token @ in JSON at position 15","code":"InvalidRequest"}",
  "headers": {
    "Content-Type": "application/json",
  },
  "statusCode": 400,
}
`;

exports[`api gateway wrapper tests should throw Error, returns results matching snapshot 3`] = `
{
  "body": "{"message":"forbidden"}",
  "headers": {
    "Content-Type": "application/json",
  },
  "statusCode": 403,
}
`;

exports[`api gateway wrapper tests should throw Error, returns results matching snapshot 4`] = `
{
  "body": "{"message":"BadRequest","code":"BadRequest"}",
  "headers": {
    "Content-Type": "application/json",
  },
  "statusCode": 400,
}
`;

exports[`api gateway wrapper tests should throw Error, returns results matching snapshot 5`] = `
{
  "body": "{"message":"Request data is invalid. Unexpected end of JSON input","code":"InvalidRequest"}",
  "headers": {
    "Content-Type": "application/json",
  },
  "statusCode": 400,
}
`;

exports[`api gateway wrapper tests should throw Error, returns results matching snapshot 6`] = `
{
  "body": "{"message":"forbidden"}",
  "headers": {
    "Content-Type": "application/json",
  },
  "statusCode": 403,
}
`;
